### PR TITLE
Setup deployment environment

### LINKED ISSUE #1: Deployment Environment
#### As a developer I want a continuous deployment environment setup for my new Rails application.

Acceptance criteria:
- [ ] [GitHub](https://github.com/AppFolioOnboarding) is used for source
  control.
- [ ] The project's version of both Ruby and Rails are used both locally and on
  Heroku.
- [ ] The landing page, `/`, provides some content.
- [ ] Every available controller action has corresponding tests.
- [ ] [CircleCI](https://circleci.com/) is used for Continuous Integration.
- [ ] Heroku is used to deploy the application.
- [ ] Pull requests are automatically deployed as review applications.
- [ ] The staging application automatically deploys whenever the master branch
  is updated.
- [ ] The staging application automatically deploys only after all tests pass
  on CircleCI.
- [ ] The staging application deployment can be promoted to the production
  application deployment.
- [ ] [Pronto](https://github.com/mmozuras/pronto) is used to help with code
  style consistency.

Discussion Topics:
- [ ] Where can we discover how long it takes to return an HTTP response?
- [ ] When should we commit our code?

References:
* https://rvm.io/rvm/install#try-out-your-new-rvm-installation
* http://guides.rubyonrails.org/getting_started.html
* http://guides.rubyonrails.org/testing.html#functional-tests-for-your-controllers
* https://devcenter.heroku.com/articles/getting-started-with-rails4
* https://devcenter.heroku.com/articles/pipelines
* https://github.com/bbatsov/ruby-style-guide

### PR DESCRIPTION
**Pull request by tlconnor**
Closes #1

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -22,3 +22,8 @@
 
 # Don't check in node_modules
 node_modules
+# Added by webpack-rails
+/node_modules
+/public/webpack
+
+.idea
```
Comment (bboe) [inline .gitignore]: We typically recommend putting editor specific ignores in your global gitignore file rather than the project one. Otherwise every project's gitignore file would be full of all the possible editor ignores.

File: Gemfile [hunk 2]
```diff
@@ -5,17 +5,12 @@ ruby '2.3.0'
 gem 'rails', '~> 5.0.0'
 # Use Puma as the app server
 gem 'puma', '~> 3.0'
-# Use SCSS for stylesheets
-gem 'sass-rails', '~> 5.0'
-# Use Uglifier as compressor for JavaScript assets
-gem 'uglifier', '>= 1.3.0'
+
+gem 'pg'
```
Comment (bboe) [inline Gemfile]: We intentionally want to use sqlite in development. While you need not stick with it, the intended purpose is to realize the databases need not be the same -- and that there are separate environments who configuration can be different.

File: Gemfile [hunk 3]
```diff
@@ -34,27 +35,18 @@ group :development, :test do
...
   gem 'listen', '~> 3.0.5'
   # Spring speeds up development by keeping your application running in the background. Read more: https://github.com/rails/spring
   gem 'spring'
   gem 'spring-watcher-listen', '~> 2.0.0'
   gem 'pronto-rubocop'
 end
 
-group :production do
-  gem 'pg'
-  gem 'rails_12factor'
```
Comment (bboe) [inline Gemfile]: Do we still need this gem for serving static assets on heroku?

File: app/components/styles.scss [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+	  margin: 0;
+	  padding: 0;
+	}
+}
+
+.app {
+	position: absolute;
+	width: 100%;
+	height: 100%;
+}
```
Comment (bboe) [inline app/components/styles.scss]: 👾  can you configure your editor to add newlines at the end of the file?

File: config/application.rb [hunk 5]
```diff
@@ -15,16 +15,23 @@ class Application < Rails::Application
...
+    end
+
     csp_settings = ["connect-src 'self'",
                     "default-src 'none'",
-                    'font-src https://cdn.jsdelivr.net',
+                    'font-src https://cdn.jsdelivr.net https://maxcdn.bootstrapcdn.com',
                     "img-src 'self' http: https:",
-                    "script-src 'self' 'unsafe-eval'",
-                    "style-src 'self' https://cdn.jsdelivr.net"]
+                    script_src,
```
Comment (bboe) [inline config/application.rb]: Is this allowing everything, or preventing everything? If the former, can you whitelist the domains you require? Same with style_src.

File: config/database.yml [hunk 6]
```diff
@@ -5,21 +5,22 @@
 #   gem 'sqlite3'
 #
 default: &default
-  adapter: sqlite3
-  pool: 5
```
Comment (bboe) [inline config/database.yml]: I'm curious. Does the `pool: 5` setting not work for the test environment?

File: config/routes.rb [hunk 7]
```diff
@@ -1,3 +1,4 @@
 Rails.application.routes.draw do
   # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
+  get '/', to: 'application#index'
```
Comment (bboe) [inline config/routes.rb]: Can you use `root` for this?

File: config/webpack.config.js [hunk 8]
```diff
@@ -0,0 +1,76 @@
+var path = require('path');
+var webpack = require('webpack');
+var StatsPlugin = require('stats-webpack-plugin');
```
Comment (bboe) [inline config/webpack.config.js]: Is this the right naming scheme for a module? It looks to be different than `path` and `webpack`.

File: lib/tasks/test.rake [hunk 9]
```diff
@@ -0,0 +1 @@
+task :test => 'webpack:compile'
```
Comment (bboe) [inline lib/tasks/test.rake]: 👾

File: test/controllers/application_controller_test.rb [hunk 10]
```diff
@@ -0,0 +1,7 @@
+class ApplicationControllerTest < ActionDispatch::IntegrationTest
+  test 'should get index' do
+    get '/'
+    assert_response :success
+    assert_template :index
```
Comment (bboe) [inline test/controllers/application_controller_test.rb]: In rails5 `assert_template` is gone. Can you use `assert_select` instead to ensure that the body of the HTTP response contains what you expect?

File: test/fixtures/images.yml [hunk 11]
```diff
@@ -0,0 +1,7 @@
+# Read about fixtures at http://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
```
Comment (bboe) [inline test/fixtures/images.yml]: We try to avoid using fixtures. The files are still needed to ensure that the table is cleared on each test run, but otherwise we try not to add anything to them.

File: test/models/image_test.rb [hunk 12]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
```
Comment (bboe) [inline test/models/image_test.rb]: Is this file needed?

**Comment by bboe** [comment]
Added my initial round of comments. This looks awesome. I am really excited to see how well this all integrates.

As we discussed before, it looks like you may have some work to do to get circle ci to test your work. When you're ready for another round of review please PM me on slack.

File: Procfile [hunk 13]
```diff
@@ -1 +1,2 @@
 web: bundle exec puma -C config/puma.rb
+webpack: ./node_modules/.bin/webpack-dev-server --config config/webpack.config.js
```
Comment (rmacklin) [inline Procfile]: Hi @tlconnor, @bboe shared this PR with me and let me you'll be doing Ropes differently.

I wanted to mention that this Procfile is used on Heroku, where the webpack dev server is irrelevant. It's probably okay because this line would be ignored (though I'm not positive), but you may want to move it to a development-only Procfile

File: .gitignore [hunk 14]
```diff
@@ -22,3 +22,8 @@
 
 # Don't check in node_modules
 node_modules
+# Added by webpack-rails
+/node_modules
```
Comment (rmacklin) [inline .gitignore]: Is this necessary given line 24?

File: Gemfile (hunk 2 above)
Comment (tlconnor) [inline Gemfile]: Makes sense.  In this case I'll just stick to PG.

File: Gemfile (hunk 3 above)
Comment (tlconnor) [inline Gemfile]: I assume you are talking about rails_12factor?  It is no longer required as of rails 5.  Perhaps this should be removed from the ropes template app?

(see Rails 5 notes at https://github.com/heroku/rails_12factor)

File: app/components/styles.scss (hunk 4 above)
Comment (tlconnor) [inline app/components/styles.scss]: Thanks, now added.  As a side note, it could be useful to add something like scss-lint which would capture these sorts of things. https://github.com/brigade/scss-lint

File: config/application.rb (hunk 5 above)
Comment (tlconnor) [inline config/application.rb]: The idea here is to allow javascript and css to be served via the webpack dev server.  I've also whitelisted the bootstrap CDN to allow the loading of bootstrap.

I forget why I added "http: https:" but it looks like they were not needed.  They are now removed.

File: config/database.yml (hunk 6 above)
Comment (tlconnor) [inline config/database.yml]: I guess it depends on the CI config.  By default the tests would be single threaded so a connection pool shouldn't be necessary.  I could be wrong though.

File: config/routes.rb (hunk 7 above)
Comment (tlconnor) [inline config/routes.rb]: Quite right, my rails rustiness is showing

File: config/webpack.config.js (hunk 8 above)
Comment (tlconnor) [inline config/webpack.config.js]: I have to admit this code is auto generated so was not written by me.

I would say the reason for the naming is that `path` and `webpack` are objects/modules while `StatsPlugin` is a class that can be instantiated.

File: lib/tasks/test.rake (hunk 9 above)
Comment (tlconnor) [inline lib/tasks/test.rake]: I have no idea what that emoji means!

The intention here is to run "webpack:compile" before "test".  Do you think there is a better way (I haven't used rake in a couple of years)?

File: Gemfile (hunk 3 above)
Comment (bboe) [inline Gemfile]: Thanks for pointing that out. I have removed it from the base project:
https://github.com/AppFolioOnboarding/base/commit/cf3fc6321561848fa629dfe3e3bda4fab5636e50

If you see anything else that may not be required, or done differently please let me know so I we can keep the base project up to date.

File: app/components/styles.scss (hunk 4 above)
Comment (bboe) [inline app/components/styles.scss]: Do you want to try to configure it for your project and I'll copy whatever you come up with over to the base project?

File: config/application.rb (hunk 5 above)
Comment (bboe) [inline config/application.rb]: Can we be more specific to whitelist just the dev server? This seems to allow JS and CSS from any domain, including inline CSS and script tags.

File: config/database.yml (hunk 6 above)
Comment (bboe) [inline config/database.yml]: I was curious if your removal from the default section was due to it breaking in the test environment.

File: config/webpack.config.js (hunk 8 above)
Comment (bboe) [inline config/webpack.config.js]: I suppose as long as the style guide says things should be done that way that's fine. It does appear to be inconsistent at first glance.

This is the style guide that I believe we try to follow for our JS code: https://github.com/airbnb/javascript

File: test/fixtures/images.yml (hunk 11 above)
Comment (tlconnor) [inline test/fixtures/images.yml]: So I did this exactly as expected? (this file was auto generated via `rails g model image`)

File: lib/tasks/test.rake (hunk 9 above)
Comment (bboe) [inline lib/tasks/test.rake]: Sorry, I used it in a earlier comment. I use 👾  (space invader) when I see white-space related issues. In this case the trailing newline is missing. Fortunately the lint tools should catch these things, but I guess rake files aren't checked.

File: test/controllers/application_controller_test.rb (hunk 10 above)
Comment (tlconnor) [inline test/controllers/application_controller_test.rb]: Thanks, fixed.

File: test/fixtures/images.yml (hunk 11 above)
Comment (bboe) [inline test/fixtures/images.yml]: Yeah, everyone runs into the same issue. We prefer a clean table to start with for all our tests and using code in the test to create the state we need. So we remove some things that rails generates for us.

File: test/models/image_test.rb (hunk 12 above)
Comment (tlconnor) [inline test/models/image_test.rb]: It was auto generated when I created the model.  I've removed it for now and will add it back if needed.

File: Procfile (hunk 13 above)
Comment (tlconnor) [inline Procfile]: Yeah I'm pretty sure Heroku will ignore the line unless you assign a dyno to it.  I'll move it to a development only Procfile if you like though as you say its probably not necessary

File: .gitignore (hunk 14 above)
Comment (tlconnor) [inline .gitignore]: You are right, it was automatically added when i added webpack-rails.  Now fixed.

File: lib/tasks/test.rake (hunk 9 above)
Comment (bboe) [inline lib/tasks/test.rake]: This newline is still missing from here. 👾

File: app/models/image.rb [hunk 15]
```diff
@@ -0,0 +1,2 @@
+class Image < ApplicationRecord
```
Comment (bboe) [inline app/models/image.rb]: I see that you've added this model and a migration, however, I don't see a `can create valid image` test for the model.

**Comment by bboe** [comment]
@tlconnor, I went through again and found two little things. It also looks like this branch needs a rebase. While you're at that I noticed that there's a lot added to the "set up deployment environment commit". Could you break that up into [atomic commits](http://www.freshconsulting.com/atomic-commits/)? If you're not familiar with the term, in essence, each commit should introduce a small, but independent change. The introduction of the model (and its test, always with the associated test) for instance is an atomic unit as I can then use that model in the console.

**Comment by rmacklin** [comment]
> [atomic commits](http://www.freshconsulting.com/atomic-commits/)

Hadn't seen that resource before, thanks for sharing. I also liked this one that was shared in the comments: https://medium.com/@fagnerbrack/one-commit-one-change-3d10b10cebbf#.7ybjh9ip3 and its sequel: https://medium.com/@fagnerbrack/one-pull-request-one-concern-e84a27dfe9f1#.w8nxj8ymh

**Comment by tlconnor** [comment]
@bboe @rmacklin I have split the commit to be more atomic.  The commit "Switch from asset pipeline to webpack" is still fairly large as I didn't find a sensible way to break it down further.

I've also removed the `Image` model as it is not relevant to this first step (it will instead be introduced in step 2)

File: Gemfile.lock [hunk 16]
```diff
@@ -135,11 +128,6 @@ GEM
       nokogiri (~> 1.6.0)
     rails-html-sanitizer (1.0.3)
       loofah (~> 2.0)
-    rails_12factor (0.0.3)
```
Comment (bboe) [inline Gemfile.lock]: I know typing is probably a bit of a challenge so I won't force you to fix this. Nevertheless, it looks like this gemfile.lock change belongs with the previous commit.

File: config/application.rb (hunk 5 above)
Comment (bboe) [inline config/application.rb]: Do you have a response to the above comment? It seems like we've lost some of the security that was previously set up.

File: config/webpack.config.js (hunk 8 above)
Comment (bboe) [inline config/webpack.config.js]: I'll ignore this for now, however, I think it would be great if there's going to be a ton of javascript to get eslint (or similar) set up to enforce a javascript style guide.

**Comment by bboe** [comment]
I agree it's hard to see where to partition that commit. So let's stick with what's there. I would like to see more precise content security policy settings to whitelist only the webpack server.

File: Gemfile.lock (hunk 16 above)
Comment (tlconnor) [inline Gemfile.lock]: ha, I was hoping you wouldn't notice this one  :)

And you are right, I should have split the Gemfile.lock changes correctly.

Comment (bboe) [inline Gemfile.lock]: 😎
